src/lib/field.ts

```diff
@@ -0,0 +1,794 @@
+    return new Field(Fp.random());
+  }
+
+  // internal stuff
+
+  // Provable<Field>
+  /**
+   * Part of the {@link Provable} interface.
+   *
+   * Returns an array containing this field element.
+   */
+  static toFields(x: Field) {
+    return [x];
+  }
+  /**
+   * Part of the {@link Provable} interface.
+   *
+   * Returns an empty array.
+   */
+  static toAuxiliary(): [] {
+    return [];
+  }
+  /**
+   * Part of the {@link Provable} interface.
+   *
+   * Returns 1.
+   */
+  static sizeInFields() {
+    return 1;
+  }
+  /**
+   * Part of the {@link Provable} interface.
+   *
+   * Deserializes a Field from a list of field elements.
+   */
+  static fromFields([x]: Field[]) {
+    return x;
+  }
+  /**
+   * Part of the {@link Provable} interface.
+   *
+   * Does nothing.
+   */
+  static check() {}
+  /**
+   * Part of the {@link Provable} interface.
+   *
+   * Returns an array containing this field element.
+   */
+  toFields() {
+    return Field.toFields(this);
+  }
+  /**
+   * Part of the {@link Provable} interface.
+   *
+   * Returns an empty array.
+   */
+  toAuxiliary() {
+    return Field.toAuxiliary();
+  }
+
+  // ProvableExtended<Field>
+  /**
+   * Serialize {@link Field} to a JSON string.
+   *
+   * This operation does NOT affect the circuit and can't be used to prove anything about the string representation of the Field.
+   */
+  toJSON() {
+    return this.toString();
+  }
+  /**
+   * Serialize a {@link Field} to a JSON string.
+   *
+   * This operation does NOT affect the circuit and can't be used to prove anything about the string representation of the Field.
+   */
+  static toJSON(x: Field) {
+    return x.toJSON();
+  }
+  /**
+   * Deserialize a JSON structure into a {@link Field}.
+   *
+   * This operation does NOT affect the circuit and can't be used to prove anything about the string representation of the Field.
+   */
+  static fromJSON(json: string) {
+    return new Field(Fp.fromJSON(json));
+  }
+  static toInput(x: Field) {
+    return { fields: [x] };
+  }
+
+  // Binable<Field>
+  static toBytes(x: Field) {
+    return FieldBinable.toBytes(x);
+  }
+  static readBytes<N extends number>(
+    bytes: number[],
+    offset: NonNegativeInteger<N>
+  ) {
+    return FieldBinable.readBytes(bytes, offset);
+  }
+  static fromBytes(bytes: number[]) {
+    return FieldBinable.fromBytes(bytes);
+  }
+  static sizeInBytes() {
+    return Fp.sizeInBytes();
+  }
+}
+
+const FieldBinable = defineBinable({
+  toBytes(t: Field) {
+    return [...t.toConstant().value[1]];
+  },
+  readBytes(bytes, offset) {
+    let uint8array = new Uint8Array(32);
+    uint8array.set(bytes.slice(offset, offset + 32));
+    return [
+      Object.assign(Object.create(new Field(1).constructor.prototype), {
+        value: [0, uint8array],
+      }) as Field,
+      offset + 32,
+    ];
+  },
+});
+
+function isField(x: unknown): x is Field {
+  return x instanceof Field || (x as any) instanceof SnarkyFieldConstructor;
```

Comment:
in some places we still have to deal with the possibility of field classes created in OCaml